### ISSUE TITLE
Assertationerror: Sanity check failed!

### ISSUE THREAD

**Issue by ravenspired** [pyinstaller/pyinstaller#6325]
Please help, I've been trying to fix this issue for the past 3 days without avail. I've reinstalled python, homebrew, and xcode.
## Description of the issue
Doing pyinstaller main.py errors out with
Traceback (most recent call last):
  File "/opt/homebrew/bin/pyinstaller", line 8, in <module>
    sys.exit(run())
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/__main__.py", line 126, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/__main__.py", line 65, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/build_main.py", line 815, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/build_main.py", line 762, in build
    exec(code, spec_namespace)
  File "/Users/anton/test/main.spec", line 23, in <module>
    exe = EXE(pyz,
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/api.py", line 524, in __init__
    self.__postinit__()
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/datastruct.py", line 159, in __postinit__
    self.assemble()
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/api.py", line 761, in assemble
    osxutils.fix_exe_for_code_signing(self.name)
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/utils/osx.py", line 211, in fix_exe_for_code_signing
    assert linkedit_seg.fileoff + linkedit_seg.filesize == \
AssertionError: Sanity check failed!

### Context information (for bug reports)

* Output of `pyinstaller --version`: ```4.5.1```
* Version of Python: Python 3.9.7
* Platform: OS X on M1 CPU
* How you installed Python: brew
* Did you also try this on another platform? Does it work there? Yes

* try the latest development version, using the following command:

```shell
pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip
```

* follow *all* the instructions in our "If Things Go Wrong" Guide
  (https://github.com/pyinstaller/pyinstaller/wiki/If-Things-Go-Wrong) and

### Make sure [everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly)

  * [x] start with clean installation
  * [x] use the latest development version
  * [x] Run your frozen program **from a command window (shell)** — instead of double-clicking on it
  * [x] Package your program in **--onedir mode**
  * [x] Package **without UPX**, say: use the option `--noupx` or set `upx=False` in your .spec-file
  * [x] Repackage you application in **verbose/debug mode**. For this, pass the option `--debug` to `pyi-makespec` or `pyinstaller` or use `EXE(..., debug=1, ...)` in your .spec file.

### A minimal example program which shows the error

```
print("hello")
```

### Stacktrace / full error message

```
23 INFO: PyInstaller: 4.5.1
23 INFO: Python: 3.9.7
30 INFO: Platform: macOS-12.0.1-arm64-arm-64bit
31 INFO: wrote /Users/anton/test/main.spec
32 INFO: UPX is not available.
33 INFO: Extending PYTHONPATH with paths
['/Users/anton/test', '/Users/anton/test']
97 INFO: checking Analysis
100 INFO: Building because /opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks/rthooks/pyi_rth_pkgutil.py changed
100 INFO: Initializing module dependency graph...
100 INFO: Caching module graph hooks...
103 INFO: Analyzing base_library.zip ...
1062 INFO: Processing pre-find module path hook distutils from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks/pre_find_module_path/hook-distutils.py'.
1062 INFO: distutils: retargeting to non-venv dir '/opt/homebrew/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9'
2230 INFO: Caching module dependency graph...
2290 INFO: running Analysis Analysis-00.toc
2293 INFO: Analyzing /Users/anton/test/main.py
2294 INFO: Processing module hooks...
2294 INFO: Loading module hook 'hook-xml.etree.cElementTree.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2294 INFO: Loading module hook 'hook-lib2to3.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2299 INFO: Loading module hook 'hook-encodings.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2327 INFO: Loading module hook 'hook-distutils.util.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2328 INFO: Loading module hook 'hook-pickle.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2329 INFO: Loading module hook 'hook-heapq.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2329 INFO: Loading module hook 'hook-difflib.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2330 INFO: Loading module hook 'hook-multiprocessing.util.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2330 INFO: Loading module hook 'hook-sysconfig.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2334 INFO: Loading module hook 'hook-xml.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2351 INFO: Loading module hook 'hook-distutils.py' from '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks'...
2359 INFO: Looking for ctypes DLLs
2366 INFO: Analyzing run-time hooks ...
2367 INFO: Including run-time hook '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks/rthooks/pyi_rth_pkgutil.py'
2369 INFO: Including run-time hook '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks/rthooks/pyi_rth_multiprocessing.py'
2370 INFO: Including run-time hook '/opt/homebrew/lib/python3.9/site-packages/PyInstaller/hooks/rthooks/pyi_rth_inspect.py'
2372 INFO: Looking for dynamic libraries
2396 INFO: Looking for eggs
2396 INFO: Using Python library /opt/homebrew/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/Python
2397 INFO: Warnings written to /Users/anton/test/build/main/warn-main.txt
2410 INFO: Graph cross-reference written to /Users/anton/test/build/main/xref-main.html
2417 INFO: checking PYZ
2418 INFO: Building because toc changed
2418 INFO: Building PYZ (ZlibArchive) /Users/anton/test/build/main/PYZ-00.pyz
2622 INFO: Building PYZ (ZlibArchive) /Users/anton/test/build/main/PYZ-00.pyz completed successfully.
2625 INFO: EXE target arch: arm64
2625 INFO: Code signing identity: None
2625 INFO: checking PKG
2625 INFO: Building because /Users/anton/test/build/main/PYZ-00.pyz changed
2625 INFO: Building PKG (CArchive) PKG-00.pkg
2630 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
2630 INFO: Bootloader /opt/homebrew/lib/python3.9/site-packages/PyInstaller/bootloader/Darwin-64bit/run
2630 INFO: checking EXE
2630 INFO: Building EXE because EXE-00.toc is non existent
2630 INFO: Building EXE from EXE-00.toc
2630 INFO: Copying bootloader exe to /Users/anton/test/build/main/main
2630 INFO: Converting EXE to target arch (arm64)
2638 INFO: Removing signature(s) from EXE
2649 INFO: Fixing EXE for code signing /Users/anton/test/build/main/main
Traceback (most recent call last):
  File "/opt/homebrew/bin/pyinstaller", line 8, in <module>
    sys.exit(run())
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/__main__.py", line 126, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/__main__.py", line 65, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/build_main.py", line 815, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/build_main.py", line 762, in build
    exec(code, spec_namespace)
  File "/Users/anton/test/main.spec", line 23, in <module>
    exe = EXE(pyz,
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/api.py", line 524, in __init__
    self.__postinit__()
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/datastruct.py", line 159, in __postinit__
    self.assemble()
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/building/api.py", line 761, in assemble
    osxutils.fix_exe_for_code_signing(self.name)
  File "/opt/homebrew/lib/python3.9/site-packages/PyInstaller/utils/osx.py", line 211, in fix_exe_for_code_signing
    assert linkedit_seg.fileoff + linkedit_seg.filesize == \
AssertionError: Sanity check failed!
```

Thank you!

**Comment by rokm** [comment]
Can you try with just-released 4.6? We've reworked that part of the code in #6188 and that sanity check if gone...

**Comment by ravenspired** [comment]
Thank you! It works now. Closing issue. By the way, do you know exactly why pyinstaller got a sanity check error?

**Comment by rokm** [comment]
> Closing issue. By the way, do you know exactly why pyinstaller got a sanity check error?

Not really, I'm afraid. 

It seems that in some cases, when we remove signatures from the bootloader executable using the `codesign` utility (so we can append embedded archive to executable), the `codesign` produces headers where the end of __LINKEDIT segment does not exactly match the end of the SYMTAB section (AFAICT, both should be aligned with the end of file) and points beyond the end of file (in my case, it way by 2 bytes off). We used to check for that in the sanity check, but seeing how it might get violated, I've changed the way we adjust the section and segment length, so that it is not so strictly tied to the original (modified) values.
